File: scripts/extract-colors.mjs
import sharp from "sharp";

async function extractColors(imagePath, n = 6) {
  const { data, info } = await sharp(imagePath)
    .resize(200, 200, { fit: "inside" })
    .raw()
    .ensureAlpha()
    .toBuffer({ resolveWithObject: true });

  const pixels = [];
  for (let i = 0; i < data.length; i += 4) {
    const r = data[i];
    const g = data[i + 1];
    const b = data[i + 2];
    const a = data[i + 3];
    if (a < 128) continue; // bỏ pixel trong suốt
    // Bỏ màu rất sáng (gần trắng) và rất tối (gần đen)
    if (r > 240 && g > 240 && b > 240) continue;
    if (r < 15 && g < 15 && b < 15) continue;
    pixels.push([r, g, b]);
  }

  // Quantize để nhóm màu gần nhau
  const step = 24;
  const map = new Map();
  for (const [r, g, b] of pixels) {
    const key = `${Math.floor(r / step) * step},${Math.floor(g / step) * step},${Math.floor(b / step) * step}`;
    map.set(key, (map.get(key) || 0) + 1);
  }

  const sorted = Array.from(map.entries()).sort((a, b) => b[1] - a[1]);
  const top = sorted.slice(0, n);

  return top.map(([key, count]) => {
    const [r, g, b] = key.split(",").map(Number);
    // Tính trung bình màu thực tế trong nhóm (đơn giản: dùng trung tâm bucket)
    const hex = `#${((1 << 24) + (r + step / 2) * 65536 + (g + step / 2) * 256 + (b + step / 2)).toString(16).slice(1)}`;
    return { hex, rgb: [Math.round(r + step / 2), Math.round(g + step / 2), Math.round(b + step / 2)], count };
  });
}

const colors = await extractColors("public/logo.png", 6);
console.log("🎨 Palette từ logo:\n");
colors.forEach((c, i) => {
  console.log(`${i + 1}. ${c.hex}  (RGB: ${c.rgb.join(",")})  – tần suất: ${c.count}`);
});

File: scripts/extract_colors.py
from PIL import Image
from collections import Counter
import sys

def get_dominant_colors(image_path, n=6):
    img = Image.open(image_path).convert("RGB")
    # Resize để xử lý nhanh hơn
    img = img.resize((200, 200))
    pixels = list(img.getdata())

    # Làm tròn màu để nhóm các sắc thái gần nhau
    def quantize(c, step=24):
        return tuple((v // step) * step for v in c)

    quantized = [quantize(p) for p in pixels]
    # Bỏ qua màu trắng và đen quá gần
    filtered = [c for c in quantized if not all(v > 240 for v in c) and not all(v < 20 for v in c)]
    if not filtered:
        filtered = quantized

    counter = Counter(filtered)
    most_common = counter.most_common(n)

    colors = []
    for (r, g, b), count in most_common:
        hex_color = "#{:02x}{:02x}{:02x}".format(r, g, b)
        colors.append({"hex": hex_color, "rgb": (r, g, b), "count": count})
    return colors

if __name__ == "__main__":
    path = sys.argv[1] if len(sys.argv) > 1 else "public/logo.png"
    print(f"Phân tích màu từ: {path}\n")
    colors = get_dominant_colors(path, n=6)
    for i, c in enumerate(colors, 1):
        r, g, b = c["rgb"]
        print(f"{i}. {c['hex']}  (RGB: {r},{g},{b})  – xuất hiện ~{c['count']} lần")
